daemon/logger/local/read.go

```diff
@@ -99,63 +98,71 @@ func getTailReader(ctx context.Context, r loggerutils.SizeReaderAt, req int) (io
 type decoder struct {
 	rdr   io.Reader
 	proto *logdriver.LogEntry
-	buf   []byte
+	// buf keeps bytes from rdr.
+	buf []byte
+	// offset is the position in buf.
+	// If offset > 0, buf[offset:] has bytes which are read but haven't used.
+	offset int
+	// nextMsgLen is the length of the next log message.
+	// If nextMsgLen = 0, a new value must be read from rdr.
+	nextMsgLen int
 }
 
-func (d *decoder) Decode() (*logger.Message, error) {
-	if d.proto == nil {
-		d.proto = &logdriver.LogEntry{}
-	} else {
-		resetProto(d.proto)
-	}
-	if d.buf == nil {
-		d.buf = make([]byte, initialBufSize)
-	}
-	var (
-		read int
-		err  error
-	)
-
+func (d *decoder) readRecord(size int) error {
+	var err error
 	for i := 0; i < maxDecodeRetry; i++ {
 		var n int
-		n, err = io.ReadFull(d.rdr, d.buf[read:encodeBinaryLen])
+		n, err = io.ReadFull(d.rdr, d.buf[d.offset:size])
+		d.offset += n
 		if err != nil {
 			if err != io.ErrUnexpectedEOF {
-				return nil, errors.Wrap(err, "error reading log message length")
+				return err
 			}
-			read += n
 			continue
 		}
-		read += n
 		break
 	}
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not read log message length: read: %d, expected: %d", read, encodeBinaryLen)
+		return err
 	}
+	d.offset = 0
+	return nil
+}
 
-	msgLen := int(binary.BigEndian.Uint32(d.buf[:read]))
-
-	if len(d.buf) < msgLen+encodeBinaryLen {
-		d.buf = make([]byte, msgLen+encodeBinaryLen)
+func (d *decoder) Decode() (*logger.Message, error) {
+	if d.proto == nil {
+		d.proto = &logdriver.LogEntry{}
 	} else {
-		if msgLen <= initialBufSize {
-			d.buf = d.buf[:initialBufSize]
-		} else {
-			d.buf = d.buf[:msgLen+encodeBinaryLen]
-		}
+		resetProto(d.proto)
+	}
+	if d.buf == nil {
+		d.buf = make([]byte, initialBufSize)
 	}
 
-	return decodeLogEntry(d.rdr, d.proto, d.buf, msgLen)
+	if d.nextMsgLen == 0 {
+		len, err := d.decodeSizeHeader()
+		if err != nil {
+			return nil, err
+		}
+		d.nextMsgLen = len
+	}
+	return d.decodeLogEntry()
 }
 
 func (d *decoder) Reset(rdr io.Reader) {
+	if d.rdr == rdr {
+		return
+	}
```

Comment:
> It looks like there are failures related to reading logs in the tests:

Before I simply removed dec.Reset() on logfile.go's fsnotify.Write case and that broke the tests.

Reply:
Which test did it broke?
Maybe there should be two methods? (dec.FileTruncated() and dec.Reset())

Reply:
The ones @samuelkarp mentioned above were broken.

```
[2021-11-24T19:29:54.941Z] === Failed
[2021-11-24T19:29:54.941Z] === FAIL: amd64.integration-cli TestDockerSuite/TestLogsAPIUntilFutureFollow (20.72s)
[2021-11-24T19:29:54.941Z]     docker_api_logs_test.go:155: timeout waiting for logs to exit
[2021-11-24T19:29:54.941Z]     --- FAIL: TestDockerSuite/TestLogsAPIUntilFutureFollow (20.72s)
[2021-11-24T19:29:54.941Z] 
[2021-11-24T19:29:54.941Z] === FAIL: amd64.integration-cli TestDockerSuite/TestLogsFollowSlowStdoutConsumer (1.39s)
[2021-11-24T19:29:54.941Z]     docker_cli_logs_test.go:246: assertion failed: 0 (actual int) != 150000 (expected int)
[2021-11-24T19:29:54.941Z]     --- FAIL: TestDockerSuite/TestLogsFollowSlowStdoutConsumer (1.39s)
[2021-11-24T19:29:54.941Z] 
[2021-11-24T19:29:54.941Z] === FAIL: amd64.integration-cli TestDockerSuite/TestLogsSinceFutureFollow (5.45s)
[2021-11-24T19:29:54.941Z]     docker_cli_logs_test.go:204: assertion failed: len(out) is 0: cannot read from empty log
[2021-11-24T19:29:54.941Z]     --- FAIL: TestDockerSuite/TestLogsSinceFutureFollow (5.45s)
```

I prefer to keep Decoder's interface as is. It takes io.Reader and doesn't have to know the source of the reader. So mentioning files there doesn't match with the rest of the interface.